src/Dhall/Parser.hs

```diff
@@ -164,59 +163,156 @@ sepBy1 p sep = do
             as <- sepBy p sep
             return (a:as)
 
-stringLiteral :: Parser Builder
-stringLiteral = Text.Parser.Token.stringLiteral <|> doubleSingleQuoteString
-
-doubleSingleQuoteString :: Parser Builder
-doubleSingleQuoteString = do
-    builder <- Text.Parser.Token.token p0
-    return (process builder)
+stringLiteral :: Show a => Parser a -> Parser (Expr Src a)
+stringLiteral embedded =
+    Text.Parser.Token.token
+        (   doubleQuoteLiteral embedded
+        <|> doubleSingleQuoteString embedded
+        )
+
+doubleQuoteLiteral :: Show a => Parser a -> Parser (Expr Src a)
+doubleQuoteLiteral embedded = do
+    _  <- Text.Parser.Char.char '"'
+    go
   where
-    process =
-          Data.Text.Lazy.Builder.fromLazyText
-        . Data.Text.Lazy.unlines
-        . trim
-        . Data.Text.Lazy.lines
-        . Data.Text.Lazy.Builder.toLazyText
-
-    trim lines_ = map (Data.Text.Lazy.drop shortestIndent) lines_
-      where
-        isEmpty = Data.Text.Lazy.all Data.Char.isSpace
+    go = go0 <|> go1 <|> go2
 
-        nonEmptyLines = filter (not . isEmpty) lines_
+    go0 = do
+        _ <- Text.Parser.Char.char '"'
+        return (TextLit mempty) 
 
-        indentLength line =
+    go1 = do
+        _ <- Text.Parser.Char.text "${"
+        a <- exprA embedded
```

Comment:
Yes, that's correct.  You can interpolate any arbitrary expression of type `Text`

The example you gave triggered a bug with the parser, though, which is that it wouldn't handle whitespace after `${`, but I just fixed that and now your example works:

```bash
$ dhall
let renderRating
    =   λ(rating : { name : Text, rating : Natural })
    →  ''
        # My thoughts about ${rating.name} (${
                Integer/show (Natural/toInteger rating.rating)
              } out of 5)
        ''
in  renderRating { name = "Dhall", rating = +6 }
<Ctrl-D>
Text

"\n# My thoughts about Dhall (6 out of 5)\n"
```